lib/segment/src/segment.rs

```diff
@@ -666,6 +686,56 @@ impl Segment {
             .collect()
     }
 
+    pub fn filtered_read_by_value_stream(
+        &self,
+        order_by: &OrderBy,
+        offset: Option<PointIdType>,
+        limit: Option<usize>,
+        filter: Option<&Filter>,
+    ) -> OperationResult<Vec<(f64, PointIdType)>> {
+        let payload_index = self.payload_index.borrow();
+
+        let numeric_index = payload_index
+            .field_indexes
+            .get(&order_by.key)
+            .and_then(|indexes| indexes.iter().find_map(|index| index.as_numeric()))
+            .ok_or_else(|| OperationError::ValidationError { description: "There is no range index for the `order_by` key, please create one to use `order_by`".to_string() })?;
+
+        let id_tracker = self.id_tracker.borrow();
+
+        let range_iter = numeric_index.stream_with_value(&order_by.as_range());
+
+        let ordered_iter =
+            OrderedByFieldIterator::new(range_iter, id_tracker.deref(), order_by.direction())
+                .skip_while(|item| match offset {
+                    Some(id_offset) => {
+                        let item_key = (item.value, item.external_id);
+                        let offset_key = (order_by.value_offset(), id_offset);
```

Comment:
this part can be pre-computer outside of the loop

Reply:
Yeah, that's right :D